### PR TITLE
Argo data tool

### PR DESCRIPTION
**Pull request by Marie59**
1st batch tool for the Earth-System subdomain ! This tool allows user to retrieve argo data in netcdf file. It will be part of the ocean part of the Earth System. I think it should go in the section Get Data of the tool pannel. 
I am not sure if I filed the .shed.yml correctly. 
What do you think of it ?

### CODE REVIEW

File: tools/ocean/argo_getdata.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
+# author: Marie Jossé
```
Comment (bgruening) [inline tools/ocean/argo_getdata.py:1]: can you use the argparse library for that script?

File: tools/ocean/argo_getdata.xml [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+            #else :
+              ''
+              ''
+            #end if 
+            '$output_argo'
+        ]]>
+    </command>
+    <inputs>
+        <param name="user" type="select" label="Which kind of user are you ?">
+            <option value="standard">🏊 standard mode simplifies the dataset, remove most of its jargon and return a priori good data</option>
```
Comment (bgruening) [inline tools/ocean/argo_getdata.xml:62]: Do those emojis work? :)

File: tools/ocean/argo_getdata.py (hunk 1 above)
Comment (Marie59) [reply tools/ocean/argo_getdata.py:1]: Yes I think I can just need to check how it works ;)

File: tools/ocean/argo_getdata.xml (hunk 2 above)
Comment (Marie59) [reply tools/ocean/argo_getdata.xml:62]: Yes they do I really like the style ;)

**Comment by Marie59** [comment]
I see that I need to add a category Earth-system somewhere with this error message `.. WARNING: Categories [['Earth-system']] unknown.`
But I search in the repo github here and I don't find where I am supposed to do that. Could you give me a hand @yvanlebras or @bgruening ?

File: tools/ocean/argo_getdata.py (hunk 1 above)
Comment (Marie59) [reply tools/ocean/argo_getdata.py:1]: Is it good now ?

**Comment by Marie59** [comment]
Weird my planemo t is working locally but it's not here...

**Comment by bgruening** [comment]
The warning about the missing category is because this category does not exist in the ToolShed. Use a different one from the TS please. https://toolshed.g2.bx.psu.edu/

**Comment by bgruening** [comment]
Are you running the tests locall with planemo test --biocontainers?

**Comment by yvanlebras** [comment]
Yes, categoris are toolshed ones. You can use the ecology one

**Comment by Marie59** [comment]
> Are you running the tests locall with planemo test --biocontainers?

I was not using the option --biocontainers so I tried with it and I do have the same error. I search and tried out some stuffs and I think my tool is searching for a "galaxy user" to run and is not able to find it (because it does not exist I think). I don't really know how to fix that. I tried to add a user in the Cheetah but I don't know how to do it and it may not be the right thing to do. 
Does one have you has an idea @bgruening and @yvanlebras ?

**Comment by Marie59** [comment]
> The warning about the missing category is because this category does not exist in the ToolShed. Use a different one from the TS please. https://toolshed.g2.bx.psu.edu/

Am I gonna be able to have this tool to one of the sections I have implemented in the earth system subdomain ? Because I don't want to be forced to add the entire set of ecology tools in the earth system ? 
Thinking about it maybe I should just add it in the 'Source Data' category ? and it will be in the 'Get Data' tool section ?

**Comment by bgruening** [comment]
ToolShed sections are independent of Galaxy tool panel sections. So don't worry about it here.

**Comment by bgruening** [comment]
I assume it is this line: https://github.com/euroargodev/argopy/blob/master/argopy/data_fetchers/erddap_data.py#L686

The function wants to get `Return the password database entry for the given numeric user ID.` not sure why this is needed. But in container settings and on HPC this is not a valid assumption.

**Comment by Marie59** [comment]
So, I need to contact the authors of erddap to ask if they can modify that ?

**Comment by bgruening** [comment]
I think so. Otherwise we need to hack around this problem on our side :(

**Comment by Marie59** [comment]
OOOH can I start my happy dance @bgruening ?

File: tools/ocean/macro.xml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            <param name="user" type="select" label="Which kin of user are you ?">
+                <option value="standard">🏊 standard mode simplifies the dataset, remove most of its jargon and return a priori good data</option>
+                <option value="research">🚣 research mode simplifies the dataset to its heart, preserving only data of the highest quality for research studies, including studies sensitive to small pressure and salinity bias </option>
+                <option value="expert">🏄 expert mode return all the Argo data, without any postprocessing</option>
+           </param>
+       </inputs>
+    </xml>
+    <xml name="argo_bibref">
+       <citations>
+            <citation type="bibtex">
```
Comment (bgruening) [inline tools/ocean/macro.xml:20]: Use type doi here

File: tools/ocean/macro.xml [hunk 4]
```diff
@@ -0,0 +1,33 @@
+<macros>
+    <token name="@VERSION@">0.0.1</token>
```
[the remark this answers is not in the document]
Comment (Marie59) [reply tools/ocean/macro.xml:2]: no  reason ;) I'll change that

File: tools/ocean/argo_getdata.xml [hunk 5]
```diff
@@ -0,0 +1,349 @@
+<tool id="argo_getdata" name="Argo data access" version="@VERSION@">
```
[the remark this answers is not in the document]
Comment (Marie59) [reply tools/ocean/argo_getdata.xml:1]: mmh what is the suffix macro ?

File: tools/ocean/argo_getdata.xml [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+        <import>macro.xml</import>
+    </macros>
+    <expand macro="argo_requirements">
+    </expand>
+    <required_files>
+        <include type="literal" path="argo_getdata.py"/>
+        <include type="literal" path="macro.xml"/>
+    </required_files>
+    <command detect_errors="exit_code"><![CDATA[ 
+        python $__tool_directory__/argo_getdata.py
```
[the remark this answers is not in the document]
Comment (Marie59) [reply tools/ocean/argo_getdata.xml:13]: single quotes everywhere ?

File: tools/ocean/argo_getdata.xml [hunk 7]
```diff
@@ -0,0 +1,349 @@
+<tool id="argo_getdata" name="Argo data access" version="@VERSION@" profile="22.05">
```
Comment (bgruening) [inline tools/ocean/argo_getdata.xml:1]: https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#tool-versions

Adding VERSION_SUFFIX as well?

File: tools/ocean/argo_getdata.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (bgruening) [reply tools/ocean/argo_getdata.xml:13]: everywhere where a path is encoded. here you need to assume that `$__tool_directory__` has an empty space in the path ... "in a bed deployment"

[the remark this answers is not in the document]
Comment (bgruening) [reply tools/ocean/argo_getdata.xml:13]:
```suggestion
        python '$__tool_directory__/argo_getdata.py'
```

File: tools/ocean/argo_getdata.xml [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+              --wmo=$selection.wmo
+            #else:
+              --wmo=$selection.wmo
+              --profile=$selection.profile
+            #end if
+            #if $dataset.data_dataset == 'bgc' :
+              --params=$dataset.params
+              --measured=$dataset.measured
+            #end if 
+            --output_argo=$output_argo
```
Comment (bgruening) [inline tools/ocean/argo_getdata.xml:34]: single quote 
```suggestion
            --output_argo='$output_argo'
```

File: tools/ocean/argo_getdata.xml [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+        ]]>
+    </command>
+    <inputs>
+        <param name="user" type="select" label="Which kind of user are you ?">
```
Comment (bgruening) [inline tools/ocean/argo_getdata.xml:38]: For the label ... do you really want to ask for "What user are you?" 

The option indicate more a question `We have preconfigured some mode of operations for you. What mode do you want to use?`

File: tools/ocean/argo_getdata.xml [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+                <param name="cardinal_2" type="float" min="-90" max="90" value="0" label="Input longitude max (+east/-west):"/>
+                <param name="cardinal_3" type="float" min="-90" max="90" value="0" label="Input latitude min (+north/-south):"/>
+                <param name="cardinal_4" type="float" min="-90" max="90" value="0" label="Input latitude max (+north/-south):"/>
+                <param name="pressure_1" type="float" value="0" label="Input pressure min (db)"/>
+                <param name="pressure_2" type="float" value="0" label="Input pressure max (db)"/>
+                <param name="date_1" type="text" optional="true" label="Input starting date" value="yyyy-mm"/>
+                <param name="date_2" type="text" optional="true" label="Input ending date" value="yyyy-mm"/>    
+            </when>
+            <when value="floats">
+                <param name="wmo" type="text" label="Input WMO number" help="if you want to enter multiple ones just separate them with a comma (example : 6902746, 6902755)"/>
```
Comment (bgruening) [inline tools/ocean/argo_getdata.xml:60]:
```suggestion
                <param name="wmo" type="text" label="Input WMO number" help="If you want to enter multiple ones just separate them with a comma (example : 6902746, 6902755)"/>
```

File: tools/ocean/argo_getdata.xml [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+            </when>
+            <when value="profiles">
+                <param name="wmo" type="integer" value="0" label="Input WMO number" />
+                <param name="profile" type="text" multiple="true" value="0" label="Input profile cycle number" help="if you want to enter multiple ones just separate them with a comma (example : 3,12)"/>
+            </when>
+        </conditional>
+        <!--<conditional name="sources">
```
Comment (bgruening) [inline tools/ocean/argo_getdata.xml:67]: should this be still expcluded?

File: tools/ocean/argo_getdata.xml [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+            <when value="gdac">
+                <param name="ftp" type="select" label="Input ftp server url">
+                    <option value="ftp://ftp.ifremer.fr/ifremer/argo">France Coriolis</option>
+                    <option value="ftp://usgodae.org/pub/outgoing/argo">US GODAE</option>
+                </param>
+            </when>
+            <when value="argovis"/>
+        </conditional>-->
+        <conditional name="dataset">
+            <param name="data_dataset" type="select" label="Which kind of dataset do you want ?" display="radio">
```
Comment (bgruening) [inline tools/ocean/argo_getdata.xml:83]: Do you need display=radio here?

radio is more a select and not a conditional imho

File: tools/ocean/argo_getdata.xml [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+                    <option value="PSAL_DATA_MODE">PSAL_DATA_MODE</option>
+                    <option value="PSAL_QC">PSAL_QC</option>
+                    <option value="TEMP">TEMP</option>
+                    <option value="TEMP_ADJUSTED">TEMP_ADJUSTED</option>
+                    <option value="TEMP_ADJUSTED_ERROR">TEMP_ADJUSTED_ERROR</option>
+                    <option value="TEMP_ADJUSTED_QC">TEMP_ADJUSTED_QC</option>
+                    <option value="TEMP_QC">TEMP_QC</option>
+                    <option value="TIME_QC">TIME_QC</option>
+                </param>
+                <param name="measured" type="select" multiple="true" label="Choose which variables cannot be NaN and must return value (you can select nothing)" optional="true">
```
Comment (bgruening) [inline tools/ocean/argo_getdata.xml:174]:
```suggestion
                <param name="measured" type="select" multiple="true" label="Choose which variables cannot be NaN and must return value" help="You can select nothing" optional="true">
```

Comment (bgruening) [reply tools/ocean/argo_getdata.xml:174]: But the help is a bit redundant, this should be obvious from the Galaxy GUI

File: tools/ocean/argo_getdata.xml [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+                    <option value="TIME_QC">TIME_QC</option>
+                </param>
+            </when>
+        </conditional>
+    </inputs>
+    <outputs>
+        <data name="output_argo" from_work_dir="argo_data.nc" format="netcdf" label="Argo data"/>
+    </outputs>
+    <tests>
+        <test expect_num_outputs="1">
```
Comment (bgruening) [inline tools/ocean/argo_getdata.xml:266]: more tests would be great, especially to utilize the different modes

File: tools/ocean/macro.xml [hunk 15]
```diff
@@ -0,0 +1,23 @@
+<macros>
+    <token name="@VERSION@">0.1.15</token>
+    <xml name="argo_requirements">
+        <requirements>
+            <requirement type="package" version="0.1.15">argopy</requirement>
```
Comment (bgruening) [inline tools/ocean/macro.xml:5]:
```suggestion
            <requirement type="package" version="@VERSION@">argopy</requirement>
```

File: tools/ocean/argo_getdata.xml (hunk 14 above)
Comment (Marie59) [reply tools/ocean/argo_getdata.xml:266]: will do !

**Comment by Marie59** [comment]
I think I took all your @bgruening  suggestion into account. Is there anything else that needs improvement ?

**Review by bgruening — APPROVED** [review]

**Comment by bgruening** [comment]
Looks great!

**Pull request merged**